src/assisted_test_infra/test_infra/utils/utils.py

```diff
@@ -617,3 +649,67 @@ def get_iso_download_path(entity_name: str):
 def get_major_minor_version(openshift_full_version: str) -> str:
     semantic_version = semver.VersionInfo.parse(openshift_full_version, optional_minor_and_patch=True)
     return f"{semantic_version.major}.{semantic_version.minor}"
+
+
+def waiter_decorator(func):
+    """decorate cluster waiters
+    When waiter called from cluster timeout exception may raise without information
+    Adding information about the timeout from job.batch/assisted-installer-controller,
+    Taking the last error log levels and append to the timeout message object.
+
+    We will be able to download kubeconfig in the next states:
+    installing, finalizing, installed, error, adding-hosts, cancelled, installing-pending-user-action
+
+    Sometimes logs from job.batch/assisted-installer-controller may return empty or failed when:
+    - In case pod/assisted-installer-controller-zzz in pending status (~10 minutes after installing phase)
+    - In case api vip address not accessible.
+
+    Current code will retry_success 3 times log assisted-installer-controller.
+    The code will ignore failures when trying to get assisted-installer-controller logs  and return
+    the timeout exception with pre-configured message
+    """
+
+    def update_kube_config(cluster_object):
+        # download kubeconfig and update /etc/hosts with api-vip host ip
+        try:
+            cluster_object.download_kubeconfig()
+            config_etc_hosts(
+                cluster_name=cluster_object.name,
+                base_dns_domain=cluster_object.get_details().base_dns_domain,
+                api_vip=cluster_object.get_details().api_vips[0].ip,
+            )
+            return True
+        except Exception:
+            log.info(f"Unable to download kubeconfig from cluster {cluster_object.name} - ignored")
+            return False
+
+    def filter_log_by(output, filter_by, repeat_count=2):
+        if output:
+            lines = output.splitlines()
+            reverse_list = [line + "\n" for line in lines[::-1] if filter_by in line]
+            str_represent = map(str, reverse_list[:repeat_count])
+            str_represent = "\n".join(str_represent)
+            log.info(f"assister-installer-controller filter: {str_represent}")
+            return str_represent
+        log.info("assister-installer-controller filter returned empty")
+        return ""
+
+    @wraps(func)
```

Comment:
do we need this decorator? From what I understand we don't manipulate the `func` directly, so their should be not problems